pallets/pallet-validator/src/lib.rs

```diff
@@ -0,0 +1,334 @@
+//! # Validator Pallet
+//!
+//! The Validator Pallet provides functionality for tfchain DAO validators.
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support;
+use frame_support::traits::Currency;
+use frame_support::{
+	dispatch::{DispatchErrorWithPostInfo, DispatchResultWithPostInfo},
+	pallet_prelude::*,
+};
+use sp_runtime::traits::StaticLookup;
+use sp_std::prelude::*;
+use substrate_validator_set;
+
+pub mod types;
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_system::pallet_prelude::*;
+	pub type BalanceOf<T> =
+		<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+	#[pallet::config]
+	pub trait Config:
+		frame_system::Config
+		+ substrate_validator_set::Config
+		+ pallet_membership::Config<pallet_membership::Instance1>
+	{
+		/// Because this pallet emits events, it depends on the runtime's definition of an event
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		type Currency: Currency<Self::AccountId>;
+		type CouncilOrigin: EnsureOrigin<<Self as frame_system::Config>::Origin>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	#[pallet::getter(fn validator_requests)]
+	pub type Validator<T: Config> =
+		StorageMap<_, Twox64Concat, T::AccountId, types::Validator<T::AccountId>>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn bonded)]
+	pub type Bonded<T: Config> = StorageMap<_, Twox64Concat, T::AccountId, T::AccountId>;
+
+	#[pallet::event]
+	#[pallet::metadata(T::AccountId = "AccountId")]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		Bonded(T::AccountId),
+		ValidatorCreated(T::AccountId, types::Validator<T::AccountId>),
+		ValidatorApproved(types::Validator<T::AccountId>),
+		ValidatorActivated(types::Validator<T::AccountId>),
+		ValidatorRemoved(types::Validator<T::AccountId>),
+		NodeValidatorChanged(T::AccountId),
+		NodeValidatorRemoved(T::AccountId),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		BadOrigin,
```

Comment:
Do we have to redefine this ourselves? There must be some Unauthorized Error in the substrate crates that is known in other tooling we use ( like polkadotjs).

Reply:
No idea where to find this actually..